### PR TITLE
refactor: using goroutines per Namespace for each resource Kind reconciliation

### PR DESCRIPTION
**Pull request by prometherion**
Partially addressing #148 since we're getting a huge performance boost on the expected reconciliation time since we're going to use several goroutines per resource Kind reconciliation (LimitRange, NetworkPolicy, ResourceQuota, etc) and per Namespace.

This will increase the amount of spawned goroutines with subsequent pressure on the memory side: from my actual test, I saw consumption of 1Gi of RAM reconciling a single Tenant collecting 200 Namespaces, not so bad since it's an edge case.

But the most important thing is that we're moving towards to address #49 too: we were doing something really stupid, triggering an update of the outer Resource Quotas per each Namespace, causing several 409 status codes and enqueuing back the Tenant several times. This was leaving enough time frame to force the Tenant Resource Quota budget but since Capsule is now faster, this should be less frequent.

### CODE REVIEW

File: controllers/tenant_controller.go [hunk 1]
```diff
@@ -292,142 +357,166 @@ func (r *TenantReconciler) syncAdditionalRoleBindings(tenant *capsulev1beta1.Ten
...
-					// sum them and get the Tenant one.
-					rql := &corev1.ResourceQuotaList{}
-					err = r.List(context.TODO(), rql, &client.ListOptions{
-						LabelSelector: labels.NewSelector().Add(*tr).Add(*ir),
-					})
-					if err != nil {
-						r.Log.Error(err, "Cannot list ResourceQuota", "tenantFilter", tr.String(), "indexFilter", ir.String())
-						return err
+			index := i
+			resourceQuota := q
```
[the remark this answers is not in the document]
Comment (prometherion) [reply controllers/tenant_controller.go:382]: Due to pointers, otherwise, the underlying goroutines started by `group.Go` would refer to the variables provided by the range.

File: controllers/tenant_controller.go [hunk 2]
```diff
@@ -577,6 +636,47 @@ func (r *TenantReconciler) syncNamespaces(tenant *capsulev1beta1.Tenant) (err er
...
+		}
+		// getting NetworkPolicy labels for the mutateFn
+		tl, err := capsulev1beta1.GetTypeLabel(&capsulev1beta1.Tenant{})
+		if err != nil {
+			return err
+		}
+		nl, err := capsulev1beta1.GetTypeLabel(&networkingv1.NetworkPolicy{})
+		if err != nil {
+			return err
+		}
```
Comment (MaxFedotov) [inline controllers/tenant_controller.go:658]: This could be moved out of the loop

Comment (prometherion) [reply controllers/tenant_controller.go:658]: We cannot, we need the index for building the label selector.

File: controllers/tenant_controller.go [hunk 3]
```diff
@@ -292,142 +357,166 @@ func (r *TenantReconciler) syncAdditionalRoleBindings(tenant *capsulev1beta1.Ten
...
-						return err
+			index := i
+			resourceQuota := q
+			g.Go(func() (scopeErr error) {
+				// Calculating the Resource Budget at Tenant scope just if this is put in place.
+				// Requirement to list ResourceQuota of the current Tenant
+				var tr *labels.Requirement
+				if tr, scopeErr = labels.NewRequirement(tenantLabel, selection.Equals, []string{tenant.Name}); scopeErr != nil {
+					r.Log.Error(scopeErr, "Cannot build ResourceQuota Tenant requirement")
+				}
```
Comment (MaxFedotov) [inline controllers/tenant_controller.go:389]: this also could be moved out of the loop

File: controllers/tenant_controller.go (hunk 2 above)
Comment (MaxFedotov) [reply controllers/tenant_controller.go:658]: i'm talking about these lines
```
		tl, err := capsulev1beta1.GetTypeLabel(&capsulev1beta1.Tenant{})
		if err != nil {
			return err
		}
		nl, err := capsulev1beta1.GetTypeLabel(&networkingv1.NetworkPolicy{})
		if err != nil {
			return err
		}
```
in all other funcs they are out of loop, as label names are static

**Comment by MaxFedotov** [comment]
and one more thing about getting tenant and type labels. 
In some funcs it is like 
```
var tenantLabel string
	if tenantLabel, err = capsulev1beta1.GetTypeLabel(&capsulev1beta1.Tenant{}); err != nil {
		return err
	}
	var typeLabel string
	if typeLabel, err = capsulev1beta1.GetTypeLabel(&corev1.ResourceQuota{}); err != nil {
		return err
	}
```

in others it is 
```
	var tl, ll string
	tl, err = capsulev1beta1.GetTypeLabel(&capsulev1beta1.Tenant{})
	if err != nil {
		return
	}
	ll, err = capsulev1beta1.GetTypeLabel(&rbacv1.RoleBinding{})
	if err != nil {
		return
	}
```
and in some it is 
```
	var tl string
	if tl, err = capsulev1beta1.GetTypeLabel(&capsulev1beta1.Tenant{}); err != nil {
		return
	}
	var ll string
	if ll, err = capsulev1beta1.GetTypeLabel(&corev1.LimitRange{}); err != nil {
		return
	}
```

let's make in constant may be?

File: controllers/tenant_controller.go (hunk 3 above)
Comment (prometherion) [reply controllers/tenant_controller.go:389]: No, this cannot be moved out: the `Requirement` is used to combine labels selection, based on the index.

**Comment by prometherion** [comment]
> WDYT about making this more constant?

Done!

**Review by MaxFedotov — APPROVED** [review]
lgtm!
